include/deal.II/matrix_free/fe_evaluation.h

```diff
@@ -8648,81 +8155,45 @@ FEFaceEvaluation<dim,
-  // an internal-data structure
-  std::array<unsigned int, VectorizedArrayType::size()> cells_            = {};
-  std::array<unsigned int, VectorizedArrayType::size()> face_no_          = {};
-  std::array<unsigned int, VectorizedArrayType::size()> face_orientation_ = {};
-
-  cells_[0]            = this->cell;
-  face_no_[0]          = this->face_no;
-  face_orientation_[0] = this->face_orientation;
-
-  if (fe_degree > -1)
+  if (this->data->data.front().fe_degree > 0 &&
+      fast_evaluation_supported(this->data->data.front().fe_degree,
+                                this->data->data.front().n_q_points_1d) &&
+      internal::FEFaceEvaluationImplGatherEvaluateSelector<
+        dim,
+        Number,
+        VectorizedArrayType>::supports(integration_flag,
+                                       *this->data,
+                                       internal::get_beginning<Number>(
+                                         destination),
+                                       this->dof_info->index_storage_variants
+                                         [this->dof_access_index][this->cell]))
     {
-      if (!internal::FEFaceEvaluationImplIntegrateScatterSelector<
-            dim,
-            Number,
-            VectorizedArrayType>::template run<fe_degree,
-                                               n_q_points_1d>(
-            n_components,
-            1,
-            internal::get_beginning<Number>(destination),
-            shared_vector_data,
-            *this->data,
-            *this->dof_info,
-            this->begin_dof_values(),
-            this->begin_values(),
-            this->begin_gradients(),
-            this->begin_hessians(),
-            this->scratch_data.begin(),
-            evaluation_flag & EvaluationFlags::values,
-            evaluation_flag & EvaluationFlags::gradients,
-            evaluation_flag & EvaluationFlags::hessians,
-            this->active_fe_index,
-            this->first_selected_component,
-            cells_,
-            face_no_,
-            this->subface_index,
-            this->dof_access_index,
-            face_orientation_,
-            this->descriptor->face_orientations))
-        {
-          // if we arrive here, writing into the destination vector did not
-          // succeed because some of the assumptions in integrate_scatter were
-          // not fulfilled (e.g. an element or degree that does not support
-          // direct writing), so we must do it here
-          this->distribute_local_to_global(destination);
-        }
+      if (fe_degree > -1)
+        internal::FEFaceEvaluationImplIntegrateScatterSelector<
+          dim,
+          Number,
+          VectorizedArrayType>::template run<fe_degree,
+                                             n_q_points_1d>(
+          n_components,
+          integration_flag,
+          internal::get_beginning<Number>(destination),
+          shared_vector_data,
+          *this);
```

Comment:
For readability, I would prefer to have brackets here, but this is probably a different topic.